numpy/compat/py3k.py

```diff
@@ -47,7 +47,10 @@ def asstr(s):
     return str(s)
 
 def isfileobj(f):
-    return isinstance(f, (io.FileIO, io.BufferedReader, io.BufferedWriter))
+    try:
+        return isinstance(f.fileno(), int)
```

Comment:
I guess `fileno()` needs to be tried to see if it is defined, but I think this needs a test and narrowig the exception.  Python defines the exception clearly here, AFAIK.

Also, I suspect this breaks e.g. if you pass in a `bz2.BZ2File`, so I would like to test that as well (and fix it if necessary).

Reply:
You are right. `bz2.BZ2File` raise error: `io.UnsupportedOperation: Seeking is only supported on files open for reading`. Looks like need `seekable`.

Reply:
I am mostly worried that the old code manages to reject file-like objects that do have a `fileno()` but that are also compressed.  I.e. for those writing directly to the `fileno()` would be wrong because that would side-step compression.  But even a compressed file-like might be seekable?

We have a problem with that in `tofile()` which fails to reject writing to such files, actually I am curious if the old `isinstance` check we had here is specific enough to get that right (at least in practice).

Reply:
```python
def isfileobj(f):
    if isinstance(f, (io.FileIO, io.BufferedReader, io.BufferedWriter)):
        try:
            return isinstance(f.fileno(), int)
        except (io.UnsupportedOperation, AttributeError, NotImplementedError):
            pass
    return False
```
This seems more stable, although it looks a bit strange.

Reply:
Yeah, but I think it makes sense at least as a minimal fix (I do think its plausible that this should be improved more, but I am not exactly sure how).

Do we really need more than `OSError` there?  Those classes have `fileno()` defined already, and as I read it are supposed to raise `OSError` when there is no valid fileno?

Reply:
Yes, `OSError` is enough.